spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/query/SqlSpannerQuery.java

```diff
@@ -63,12 +89,68 @@ public SqlSpannerQuery(Class type, QueryMethod queryMethod,
 
 	@Override
 	public Object execute(Object[] parameters) {
-		return this.spannerOperations.find(this.entityType, SpannerStatementQueryExecutor
-				.buildStatementFromSqlWithArgs(this.sql, this.tags, parameters));
+		return this.spannerOperations.find(this.entityType,
+				SpannerStatementQueryExecutor.buildStatementFromSqlWithArgs(
+						getSql(parameters), this.tags,
+						filterOutSpelParameters(parameters)));
+	}
+
+	@VisibleForTesting
+	String getSql(Object[] parameterValues) {
+		EvaluationContext context = this.evaluationContextProvider
+				.getEvaluationContext(this.queryMethod.getParameters(), parameterValues);
+		return this.sqlExpression == null ? this.sql
+				: this.sqlExpression.getValue(context, String.class);
 	}
 
 	@Override
 	public QueryMethod getQueryMethod() {
 		return this.queryMethod;
 	}
+
+	@VisibleForTesting
+	Object[] filterOutSpelParameters(Object[] rawParams) {
```

Comment:
I looked at this for quite some time now ... I don't understand what you are doing here.

The method name suggests a relation to SpELs but you are actually checking for the presence of an `@Param` annotation, which is just a way to specify a name for a parameter.

Reply:
The issue this one was meant to address was that the SQL string might have some arguments provided by SpEL, which would be annotated with `@Param`, and then the other arguments in that SQL would be provided by the Spanner-style tags. We're filtering out those annotated for SpEL so the Spanner tags don't use the args meant for SpEL.

However, given the other concern regarding escaping/injection and the complexity of it all, I think we should not use SpEL params in these query methods. We should keep SpEL in those SQL strings for purposes of varying table names and other structural things, but not from the Params of the method.  For the args of the SQL i think we should just stick to the Spanner tags because all the escaping is taken care of for us. 

@balopat what do you think

Reply:
I think we have to significantly rework this, I'll comment in a second with more details.